devtools/cmd/migrate-librarian/main.go

```diff
@@ -0,0 +1,202 @@
+
+	if err := yaml.Write(*outputPath, cfg); err != nil {
+		return fmt.Errorf("failed to write config: %w", err)
+	}
+	slog.Info("Wrote config to output file", "path", outputPath)
+
+	if err := librarian.RunTidy(); err != nil {
+		slog.Error(errTidyFailed.Error(), "error", err)
+		return errTidyFailed
+	}
+
+	return nil
+}
+
+func buildConfig(
+	librarianState *legacyconfig.LibrarianState,
+	librarianConfig *legacyconfig.LibrarianConfig,
+	lang string) *config.Config {
+	repo := "googleapis/google-cloud-go"
+	if lang == "python" {
+		repo = "googleapis/google-cloud-python"
+	}
+
+	cfg := &config.Config{
+		Language: lang,
+		Repo:     repo,
+		Default: &config.Default{
+			TagFormat: defaultTagFormat,
+		},
+	}
+
+	libraries, sha := buildLibraries(librarianState, librarianConfig)
+	if sha != "" {
+		cfg.Sources = &config.Sources{
+			Googleapis: &config.Source{
+				Commit: sha,
+			},
+		}
+	}
+	cfg.Libraries = libraries
+
+	return cfg
+}
+
+func buildLibraries(
+	librarianState *legacyconfig.LibrarianState,
+	librarianConfig *legacyconfig.LibrarianConfig) ([]*config.Library, string) {
+	var libraries []*config.Library
+	idToLibraryState := sliceToMap[legacyconfig.LibraryState](
+		librarianState.Libraries,
+		func(lib *legacyconfig.LibraryState) string {
+			return lib.ID
+		})
+	idToLibraryConfig := sliceToMap[legacyconfig.LibraryConfig](
+		librarianConfig.Libraries,
+		func(lib *legacyconfig.LibraryConfig) string {
+			return lib.LibraryID
+		})
+	googleapisCommitSHA := ""
+	// Iterate libraries from idToLibraryState because librarianConfig.Libraries is a
+	// subset of librarianState.Libraries.
+	for id, libState := range idToLibraryState {
+		library := &config.Library{}
+		library.Name = id
+		library.Version = libState.Version
+		if libState.APIs != nil {
+			library.Channels = toChannels(libState.APIs)
+		}
+		library.Keep = libState.PreserveRegex
+		// Go and Python monorepo only contains GAPIC libraries.
+		library.SpecificationFormat = "protobuf"
+		if googleapisCommitSHA == "" {
+			googleapisCommitSHA = libState.LastGeneratedCommit
+		}
+		// hydrate params from library config
+		libCfg, ok := idToLibraryConfig[id]
+		if ok {
+			library.SkipGenerate = libCfg.GenerateBlocked
+			library.SkipRelease = libCfg.ReleaseBlocked
+		}
+
+		libraries = append(libraries, library)
+	}
+
+	sort.Slice(libraries, func(i, j int) bool {
+		return libraries[i].Name < libraries[j].Name
+	})
+
+	return libraries, googleapisCommitSHA
+}
+
+func sliceToMap[T any](slice []*T, keyFunc func(t *T) string) map[string]*T {
+	res := make(map[string]*T, len(slice))
+	for _, t := range slice {
+		key := keyFunc(t)
+		res[key] = t
+	}
+
+	return res
+}
+
+func toChannels(apis []*legacyconfig.API) []*config.Channel {
+	channels := make([]*config.Channel, 0, len(apis))
+	for _, api := range apis {
+		channels = append(channels, &config.Channel{
+			Path:          api.Path,
+			ServiceConfig: api.ServiceConfig,
```

Comment:
If https://pkg.go.dev/github.com/googleapis/librarian@main/internal/serviceconfig#Find returns a non-empty string, we don't need to list it here